pkg/console/subresource/route/route.go

```diff
@@ -35,6 +36,137 @@ type CustomTLSCert struct {
 	Key         string
 }
 
+type RouteConfig struct {
+	defaultRoute RouteControllerSpec
+	customRoute  RouteControllerSpec
+	domain       string
+	routeName    string
+}
+
+type RouteControllerSpec struct {
+	hostname   string
+	secretName string
+}
+
+func getComponentRouteSpec(ingressConfig *configv1.Ingress, componentName string) *configv1.ComponentRouteSpec {
+	for i, componentRoute := range ingressConfig.Spec.ComponentRoutes {
+		if componentRoute.Name == componentName && componentRoute.Namespace == api.OpenShiftConsoleNamespace {
+			componentRoute := ingressConfig.Spec.ComponentRoutes[i].DeepCopy()
+			return componentRoute
+		}
+	}
+	return nil
+}
+
+func getComponentRouteStatus(ingressConfig *configv1.Ingress, componentName string) *configv1.ComponentRouteStatus {
+	for i, componentRoute := range ingressConfig.Status.ComponentRoutes {
+		if componentRoute.Name == componentName && componentRoute.Namespace == api.OpenShiftConsoleNamespace {
+			componentRoute := ingressConfig.Status.ComponentRoutes[i].DeepCopy()
+			return componentRoute
+		}
+	}
+	return nil
+}
+
+func NewRouteConfig(operatorConfig *operatorv1.Console, ingressConfig *configv1.Ingress, routeName string) *RouteConfig {
+	defaultRoute := RouteControllerSpec{
+		hostname: GetDefaultRouteHost(routeName, ingressConfig),
+	}
+	var customRoute RouteControllerSpec
+	var isIngressConfigCustomHostnameSet bool
+
+	// Custom hostname in ingress config takes precedent over console operator's config
+	componentRouteSpec := getComponentRouteSpec(ingressConfig, routeName)
+	if componentRouteSpec != nil {
+		customRoute.hostname = string(componentRouteSpec.Hostname)
+		if len(componentRouteSpec.ServingCertKeyPairSecret.Name) != 0 {
+			customRoute.secretName = componentRouteSpec.ServingCertKeyPairSecret.Name
+		}
+		isIngressConfigCustomHostnameSet = true
+	}
+
+	// Legacy behaviour, operatorConfig only configures custom hostname for "console" route.
+	// The custom route hostname doesn't have to be set if admin just wants to set the custom
+	// TLS for the default route.
```

Comment:
I'm a little unclear on this comment. Are both these scenarios supported in the new API? (Setting hostname without certs or setting certs without hostname?)

Reply:
@spadgett it's up to each component how they implement the logic. Since we already had the functionality I've kept it for the console-operator.